Zend/zend_ini.c

```diff
@@ -536,6 +541,202 @@ ZEND_API bool zend_ini_parse_bool(zend_string *str)
+
+		*errstr = zend_strpprintf(0, "Invalid quantity \"%s\": no valid leading digits, interpreting as \"0\" for backwards compatibility",
+						ZSTR_VAL(invalid.s));
+
+		smart_str_free(&invalid);
+		return 0;
+	}
+
+	/* Allow for whitespace between integer portion and any suffix character */
+	while (digits_end < str_end && zend_is_whitespace(*digits_end)) ++digits_end;
+
+	/* No exponent suffix. */
+	if (digits_end == str_end) {
+		goto end;
+	}
+
+	switch (*(str_end-1)) {
+		case 'g':
+		case 'G':
+			factor = 1<<30;
+			break;
+		case 'm':
+		case 'M':
+			factor = 1<<20;
+			break;
+		case 'k':
+		case 'K':
+			factor = 1<<10;
+			break;
+		default:
+			/* Unknown suffix */
+			smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+			smart_str_0(&invalid);
+			smart_str_append_escaped(&interpreted, str, digits_end - str);
+			smart_str_0(&interpreted);
+			smart_str_append_escaped(&chr, str_end-1, 1);
+			smart_str_0(&chr);
+
+			*errstr = zend_strpprintf(0, "Invalid quantity \"%s\": unknown multipler \"%s\", interpreting as \"%s\" for backwards compatibility",
+						ZSTR_VAL(invalid.s), ZSTR_VAL(chr.s), ZSTR_VAL(interpreted.s));
+
+			smart_str_free(&invalid);
+			smart_str_free(&interpreted);
+			smart_str_free(&chr);
+
+			return retval;
+	}
+
+	if (!overflow) {
+		if (signed_result) {
+			zend_long sretval = (zend_long)retval;
+			if (sretval > 0) {
+				overflow = (zend_long)retval > ZEND_LONG_MAX / (zend_long)factor;
+			} else {
+				overflow = (zend_long)retval < ZEND_LONG_MIN / (zend_long)factor;
+			}
+		} else {
+			overflow = retval > ZEND_ULONG_MAX / factor;
+		}
+	}
+
+	retval *= factor;
+
+	if (UNEXPECTED(digits_end != str_end-1)) {
+		/* More than one character in suffix */
+		smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+		smart_str_0(&invalid);
+		smart_str_append_escaped(&interpreted, str, digits_end - str);
+		smart_str_0(&interpreted);
+		smart_str_append_escaped(&chr, str_end-1, 1);
+		smart_str_0(&chr);
+
+		*errstr = zend_strpprintf(0, "Invalid quantity \"%s\", interpreting as \"%s%s\" for backwards compatibility",
+						ZSTR_VAL(invalid.s), ZSTR_VAL(interpreted.s), ZSTR_VAL(chr.s));
+
+		smart_str_free(&invalid);
+		smart_str_free(&interpreted);
+		smart_str_free(&chr);
+
+		return (zend_long) retval;
+	}
+
+end:
+	if (UNEXPECTED(overflow)) {
+		smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+		smart_str_0(&invalid);
+
+		/* Not specifying the resulting value here because the caller may make
+		 * additional conversions. Not specifying the allowed range
+		 * because the caller may do narrower range checks. */
+		*errstr = zend_strpprintf(0, "Invalid quantity \"%s\": value is out of range, using overflow result for backwards compatibility",
+						ZSTR_VAL(invalid.s));
+
+		smart_str_free(&invalid);
+		smart_str_free(&interpreted);
+		smart_str_free(&chr);
+
+		return (zend_long) retval;
+	}
+
+	*errstr = NULL;
+	return (zend_long) retval;
```

Comment:
Converting unsigned numbers to signed numbers when out-of-range is implementation defined. Do all architectures supported by PHP agree on this behavior?

Reply:
I think you are right that it's UB. The behavior of implicit conversions and casts is defined by the same notion of conversion.

I would expect that all architectures/compilers supported by PHP do agree on this behavior. If my understanding is correct, in twos complement architectures this behavior is obtained by not changing the representation / doing nothing, so I would expect that this is what compilers do. The behavior is documented at least in GCC. An other consideration in favor of this is that we depend on it in a few places, including in zend_atol().

I've removed this conversion anyway, so that the function is UB-free at least in the zend_ini_parse_uquantity() variant :)

Reply:
The behavior is *implementation defined*, not *undefined*. We absolutely need to avoid the latter, but relying on the former can be okay. It is in this case (conversion/cast from unsigned to signed).